torch/distributed/pipelining/stage.py

```diff
@@ -1293,22 +1314,102 @@ def stage_global_rank(peer_rank):
+    ):
+        if kwargs is None:
+            kwargs = {}
+        assert args is not None, "Args may be an empty list but not None"
+        if (
+            self.is_first
+            or self.stage_index_to_group_rank[self.stage_index - 1] == self.group_rank
+        ):
+            args = tree_map_only(torch.Tensor, lambda x: x.to("meta"), args)
+        else:
+            assert (
+                len(args) == 0
+            ), "Can't supply input args for shape inference on non-first stage"
+            objects = [None]
+            logger.debug(
+                "%s receiving from stage %s", self.stage_index, self.stage_index - 1
+            )
+            dist.recv_object_list(
+                objects, src=self.prev_stage, group=self.group, device=self.device
+            )
+            recv_args = objects[0]
+            assert isinstance(recv_args, tuple), type(recv_args)
+            args = recv_args
+
+        # cache input shapes for use during recv buffer allocation
+        self.inputs_meta = args
+        args = tree_map_only(
+            torch.Tensor, lambda x: torch.zeros_like(x, device=self.device), args
+        )
+
+        # set attributes needed for forward
+        with torch.no_grad():
+            logger.debug("%s running forward", self.stage_index)
+            outputs = self.submod(*args, **kwargs)
+
+        # if single tensor, convert so it is always a list
+        if isinstance(outputs, torch.Tensor):
+            outputs = [outputs]
+
+        # communicate meta outputs not real outputs for two reasons
+        # 1 - its faster (esp. since obj coll pickles tensor data!)
+        # 2 - avoid activating a cuda context for the src rank when unpickling on the recv end!
+        outputs_meta = tuple(
+            tree_map_only(torch.Tensor, lambda x: x.to("meta"), outputs)
         )
+        self._configure_outputs_meta(outputs_meta)
+
+        if (
+            self.is_last
+            or self.stage_index_to_group_rank[self.stage_index + 1] == self.group_rank
+        ):
+            logger.debug("%s skipping send to next stage", self.stage_index)
+
+        else:
+            logger.debug(
+                "%s sending to stage %s", self.stage_index, self.stage_index + 1
+            )
+            dist.send_object_list(
+                [outputs_meta],
+                dst=self.next_stage,
+                group=self.group,
+                device=self.device,
+            )
+            outputs_meta = tuple()
+
+        return outputs_meta
 
     def _prepare_forward_infra(
         self,
         num_microbatches: int,
-    ) -> None:
+        args: Tuple[Any, ...],
+        kwargs: Optional[Dict[str, Any]] = None,
+    ) -> Tuple[Any, ...]:
         # TODO move self.device to an argument from step API (from its input tensors)?
+        assert num_microbatches is not None, "TODO fix num_microbatches"
 
+        outputs: Tuple[Any, ...] = tuple()
+        if self.inputs_meta is None:
```

Comment:
Hmm, would this `if` causes unmatched behaviors between ranks?

Reply:
yea, I am making an assumption here that every rank either has shapes or doesn't have shapes provided for the Stage ctor, but we don't have enforcement. 

i guess we could do a synchronization to agree on whether all ranks are doing shape inference before we do it.  But that would lead to another question of when- (a) in __init__ feels bad to me (i never like to do comm ops in init) (b) but at step() time we'd get stuck doing it on every step which is too expensive, or else we'd need the 'if initialized' logic to skip it, which in theory could desync.